### PR TITLE
[FLINK-25167][API / DataStream]Expose StreamOperatorFactory in Connec…

### PR DESCRIPTION
**Pull request by shouweikun**
…tedStreams#transform

## What is the purpose of the change

As title says, expose `StreamOperatorFactory` in `ConnectedStreams`#`transform`

## Brief change log
  - *Expose StreamOperatorFactory in both Java and Scala `ConnectedStreams`*

## Verifying this change

This change is a trivial code change without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (*yes*)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)

### CODE REVIEW

**Comment by shouweikun** [comment]
@flinkbot run azure

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/TwoInputStreamOperatorFactory.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+import org.apache.flink.annotation.PublicEvolving;
+
+/**
+ * A factory to create {@link TwoInputStreamOperator}.
+ *
+ * @param <IN1> The input type of the operator
+ * @param <IN2> The input type of the operator
+ * @param <OUT> The output type of the operator
+ */
+@PublicEvolving
```
Comment (fapaul) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/TwoInputStreamOperatorFactory.java:30]: It should not be possible to extend an `@Experimental` component with a higher stability guarantee. Currently, the project misses a test to cover that behavior.
Unfortunately, this means we need to make the `StreamOperatorFactory` `@PublicEvolving`.

@pnowojski wdyt?

**Review by fapaul — COMMENTED** [review]
Looks code code-wise but I think we have a stability problem with the annotations.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/TwoInputStreamOperatorFactory.java (hunk 1 above)
Comment (pnowojski) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/TwoInputStreamOperatorFactory.java:30]: +1 from my side for this. It's been `@Experimental` for long enough.

Note that just as for `StreamOperator`, for `StreamOperatorFactory` we will never intend to make it fully `@Public`.

Comment (shouweikun) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/TwoInputStreamOperatorFactory.java:30]: @fapaul Thanks for your review. As for this pr, shall we just remove `@PublicEvolving` from `TwoInputStreamOperatorFactory` and `ConnectedStreams`#`transform`(xx, xx, `TwoInputStreamOperatorFactory`)?

After checking [code](https://github.com/apache/flink/blob/d7cf2c10f8d4fba81173854cbd8be27c657c7c7f/flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/DataStream.java#L1180), I think  it has the same stability issue too. 

> Currently, the project misses a test to cover that behavior.

As you mentioned above.  Maybe we can create a new jira ticket to address this separately??

Comment (fapaul) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/TwoInputStreamOperatorFactory.java:30]: @shouweikun I'd propose marking `StreamOperatorFactory` as `@PublicEvolving` but please do it as a separate commit. All your other changes are good as they are.

**Comment by shouweikun** [comment]
Hi, @fapaul @pnowojski 
I‘d say that I encountered some problems. I tried to mark `StreamOperatorFactory`  and `StreamOperatorParameters` as  `@PublicEvoliving`, tests still failed. I checked the cause and intended to fix them but I didn't know what to do.
Asking for help.

**Comment by fapaul** [comment]
I had an internal discussion with @pnowojski and we think the best way is to remove this commit ae038db and add the new violation to the whitelist. You can do that by uncommenting this line [1] and running the ArchUnitTests [2] again locally. You also need to commit the file in the violations folder afterward.
Please ensure that you have rebased your branch to the latest master **before** running the tests.

[1] https://github.com/apache/flink/blob/06192741d02714499c6d0f671b21bfa2588a21bf/flink-architecture-tests/src/test/resources/archunit.properties#L26
[2] https://github.com/apache/flink/blob/master/flink-architecture-tests/src/test/java/org/apache/flink/architecture/ArchitectureTest.java

**Review by fapaul — APPROVED** [review]
@shouweikun LGTM, thanks for your patience. Can you squash the commits together so that in the end the PR consists of two commits with commit formatting i.e. [FLINK-XXXX][...]? 

1. Making the StreamOperatorFactory `@PublicEvolving` + the new violation
2. Adding your initial change of this PR.

**Comment by shouweikun** [comment]
@fapaul Thanks for your patience too. Comments addressed.

**Pull request merged**